proxy/coco/src/state.rs

```diff
@@ -342,30 +343,111 @@ impl State {
+        &self,
+        reference: NamespacedRef<Single>,
+        callback: F,
+    ) -> Result<T, Error>
     where
-        F: Send + FnOnce(&mut git::Browser) -> Result<T, source::Error>,
+        F: FnOnce(&mut git::Browser) -> Result<T, source::Error> + Send,
     {
+        let namespace = git::Namespace::try_from(reference.namespace().to_string().as_str())
+            .map_err(source::Error::from)?;
+        let branch = match reference.remote {
+            None => git::Branch::local(&reference.name),
+            Some(peer) => {
+                git::Branch::remote(&format!("heads/{}", reference.name), &peer.to_string())
+            },
+        };
         let monorepo = self.monorepo();
-        let project = self.get_project(urn, None).await?;
-        let default_branch = git::Branch::local(project.default_branch());
-
         let repo = git::Repository::new(monorepo).map_err(source::Error::from)?;
-        let namespace = git::Namespace::try_from(project.urn().id.to_string().as_str())
-            .map_err(source::Error::from)?;
-        let mut browser = git::Browser::new_with_namespace(&repo, &namespace, default_branch)
+        let mut browser = git::Browser::new_with_namespace(&repo, &namespace, branch)
             .map_err(source::Error::from)?;
 
         callback(&mut browser).map_err(Error::from)
     }
 
+    /// This method helps us get a branch for a given [`RadUrn`] and optional [`PeerId`].
+    ///
+    /// If the `branch_name` is `None` then we get the project for the given [`RadUrn`] and use its
+    /// `default_branch`.
+    ///
+    /// # Errors
+    ///   * If the storage operations fail.
+    ///   * If the requested reference was not found.
+    pub async fn get_branch<P, B>(
+        &self,
+        urn: RadUrn,
+        remote: P,
+        branch_name: B,
+    ) -> Result<NamespacedRef<Single>, Error>
+    where
+        P: Into<Option<PeerId>> + Clone + Send,
+        B: Into<Option<String>> + Clone + Send,
+    {
+        let name = match branch_name.into() {
+            None => {
+                let project = self.get_project(urn.clone(), None).await?;
+                project.default_branch().to_owned()
+            },
+            Some(name) => name,
+        };
+        let reference = NamespacedRef::head(urn.id, remote.clone(), &name);
+
+        let exists = {
+            let reference = reference.clone();
+            self.api
+                .with_storage(move |storage| storage.has_ref(&reference))
+                .await??
+        };
+
+        if exists {
+            Ok(reference)
+        } else {
+            Err(Error::MissingRef { reference })
+        }
+    }
+
+    /// This method helps us get the default branch for a given [`RadUrn`].
+    ///
+    /// It does this by:
+    ///     * First checking if the owner of this storage has a reference to the default
+    /// branch.
+    ///     * If the owner does not have this reference then it falls back to the first maintainer.
```

Comment:
Wonder if this leads to unexpected behaviour from the users perspective.

Reply:
Ya, I think I'll create a follow-up ticket for this because I mostly "guessed" which way of getting a branch is correct. But I'd like to audit and understand the cases better.